velox/type/Type.h

```diff
@@ -435,6 +435,37 @@ struct TypeFactory;
     return this->kind() == TypeKind::KIND;                                \
   }
 
+class Type;
+
+enum class TypeParameterKind {
+  /// Type. For example, element type in the array type.
+  kType,
+  /// Integer. For example, precision in a decimal type.
+  kLongLiteral,
+};
+
+struct TypeParameter {
+  const TypeParameterKind kind;
+
+  /// Must be not not null when kind is kType. All other properties should be
+  /// null or unset.
+  const std::shared_ptr<const Type> type;
+
+  /// Must be set when kind is kLongLiteral. All other properties should be null
+  /// or unset.
+  const std::optional<int64_t> longLiteral;
```

Comment:
I'm not sure if I'm following the idea correctly, but should this be handled through specialization instead (e.g: TypeParameterType and TypeParameterLongLiteral)?